### ISSUE TITLE
Issues with views that utilise union all

### ISSUE THREAD

**Issue by MathiasVersichele** [citusdata/citus#4144]
I'm having lots of trouble with views that utilise "union all". This is the view:

```
CREATE OR REPLACE VIEW rvpezocoav as
SELECT rtpezocoav.id,
    rtpezocoav.ad_id,
    rtpezocoav.pe_id,
    rtpezocoav.zo_id,
    rtpezocoav.co_id,
    rtpezocoav.value_num,
    rtpezocoav.value_text
   FROM rtpezocoav
     JOIN rmv_active_peco act ON act.co_id = rtpezocoav.co_id AND act.pe_id = rtpezocoav.pe_id
UNION ALL
 SELECT av.id,
    av.ad_id,
    rtpe.id AS pe_id,
    av.zo_id,
    av.co_id,
    av.value_num,
    av.value_text
   FROM rtpezocoav av
     JOIN rtpe ON av.pe_id = rtpe.baseline_id
     JOIN rmv_active_peco act ON act.co_id = av.co_id AND act.pe_id = rtpe.id
  WHERE NOT (EXISTS ( SELECT 1
           FROM rtpezocoav av_1
          WHERE av_1.pe_id = rtpe.id AND av_1.zo_id = av.zo_id AND av_1.co_id = av.co_id AND av_1.ad_id = av.ad_id))
```

- rtpezocoav is a distributed table on zo_id
- rmv_active_peco & rtpe are reference tables

The following query fails after a while with the 'intermediate result size exceeds citus.max_intermediate_result_size (currently 1048576 kB)':
```
select count(1) from rvpezocoav where pe_id = 391;
```

I have no idea why we would need to reshuffle data as it's a join between one distributed table and two reference tables.

What I tried and didn't work:
- remove join with rmv_active_peco
- remove where not (exists ...) clause

Only thing that works is changing the union ALL to just a union. No idea why that makes a difference...

But then I end up with the next problem:

```
select count(1)
	from rvpezocoav pezoco
	join rtco co on pezoco.co_id = co.lo_id
	where ad_id = 4
	        and pe_id = 391
	        and lo_id = 505
```
fails with ERROR: complex joins are only supported when all distributed tables are co-located and joined on their distribution columns.

rtco here is also a reference table.

What I tried according to suggestions on slack:

```
 CREATE OR REPLACE VIEW rvpezocoav as
				select * from 
				(
				select * from 
				(
				SELECT rtpezocoav.id,
				    rtpezocoav.ad_id,
				    rtpezocoav.pe_id,
				    rtpezocoav.zo_id,
				    rtpezocoav.co_id,
				    rtpezocoav.value_num,
				    rtpezocoav.value_text
				   FROM rtpezocoav
				     JOIN rmv_active_peco act ON act.co_id = rtpezocoav.co_id AND act.pe_id = rtpezocoav.pe_id
				     offset 0
				    ) as z1
				UNION --ALL  --!!!!!
				 select * from 
				 (
					SELECT av.id,
				    av.ad_id,
				    rtpe.id AS pe_id,
				    av.zo_id,
				    av.co_id,
				    av.value_num,
				    av.value_text
				   FROM rtpezocoav av
				     JOIN rtpe ON av.pe_id = rtpe.baseline_id
				     JOIN rmv_active_peco act ON act.co_id = av.co_id AND act.pe_id = rtpe.id
				  WHERE NOT (EXISTS ( SELECT 1
				           FROM rtpezocoav av_1
				          WHERE av_1.pe_id = rtpe.id AND av_1.zo_id = av.zo_id AND av_1.co_id = av.co_id AND av_1.ad_id = av.ad_id))
				          offset 0
				      ) as z2
				  ) as z;
```

This brings me back to the original problem: ERROR: the intermediate result size exceeds citus.max_intermediate_result_size (currently 1048576 kB)

I've been going in circles trying to make this work for the last few days... This is a real showstopper for us unfortunately.

**Comment by MathiasVersichele** [comment]
I'll work out a MWE to take it further from there.

**Comment by MathiasVersichele** [comment]
Here's the MWE. Unfortunately for me, I can't reproduce error in this MWE. But when I change the UNION ALL to UNION, I can reproduce an error (ERROR: complex joins are only supported when all distributed tables are co-located and joined on their distribution columns). Maybe that can also be handy to look at.

```
drop schema if exists test cascade;
create schema test;
select run_command_on_workers('drop schema if exists test cascade;');
select run_command_on_workers('create schema test;');


set search_path = test;

CREATE TABLE test.rtpezocoav (
	id int8 NULL,
	pe_id int4 NOT NULL,
	zo_id int4 NOT NULL,
	co_id int4 NOT NULL,
	value_num float8 NULL,
	value_text varchar NULL,
	ad_id int4 NOT NULL
);

CREATE UNIQUE INDEX rtpezocoav_unique ON test.rtpezocoav USING btree (pe_id, zo_id, co_id, ad_id);
select create_distributed_table('test.rtpezocoav', 'zo_id');

insert into test.rtpezocoav (id, ad_id, pe_id, zo_id, co_id, value_num)
select 1, 1, 1, 1, 1, 1.0;

insert into test.rtpezocoav (id, ad_id, pe_id, zo_id, co_id, value_num)
select 1, 1, 1, 2, 2, 1.0;




drop table if exists test.rtpe;
create table test.rtpe (
	id serial NOT NULL,
	"name" varchar NOT NULL,
	baseline_id int,
	CONSTRAINT rtpe_pkey PRIMARY KEY (id)
);
insert into test.rtpe(id, name, baseline_id)
select 1, 'testperiod', null;

select create_reference_table('test.rtpe');




CREATE TABLE test.rtco (
	lo_id int4 NULL,
	br_id int4 NULL,
	ch_id int4 NULL
);
CREATE INDEX rtco_br_id_idx ON test.rtco USING btree (br_id);
CREATE INDEX rtco_ch_id_idx ON test.rtco USING btree (ch_id);
CREATE UNIQUE INDEX rtco_pkey ON test.rtco USING btree (lo_id);

insert into test.rtco(lo_id, br_id, ch_id)
select 1, 1, 1;
insert into test.rtco(lo_id, br_id, ch_id)
select 2, 1, 1;

select create_reference_table('test.rtco');




CREATE TABLE test.rmv_active_peco (
	pe_id int4 NULL,
	co_id int4 NULL
);
CREATE UNIQUE INDEX rmv_active_peco_2_pe_id_co_id_idx ON test.rmv_active_peco USING btree (pe_id, co_id);

select create_reference_table('test.rmv_active_peco');

insert into test.rmv_active_peco(pe_id, co_id)
select 1, 1;



CREATE OR REPLACE VIEW test.rvpezocoav as
SELECT rtpezocoav.id,
    rtpezocoav.ad_id,
    rtpezocoav.pe_id,
    rtpezocoav.zo_id,
    rtpezocoav.co_id,
    rtpezocoav.value_num,
    rtpezocoav.value_text
   FROM test.rtpezocoav
   JOIN test.rmv_active_peco act ON act.co_id = rtpezocoav.co_id AND act.pe_id = rtpezocoav.pe_id
UNION ALL  --!!!!!
 SELECT av.id,
    av.ad_id,
    rtpe.id AS pe_id,
    av.zo_id,
    av.co_id,
    av.value_num,
    av.value_text
   FROM test.rtpezocoav av
     JOIN test.rtpe ON av.pe_id = rtpe.baseline_id
     JOIN test.rmv_active_peco act ON act.co_id = av.co_id AND act.pe_id = rtpe.id
  WHERE NOT (EXISTS ( SELECT 1
           FROM test.rtpezocoav av_1
          WHERE av_1.pe_id = rtpe.id AND av_1.zo_id = av.zo_id AND av_1.co_id = av.co_id AND av_1.ad_id = av.ad_id));

         

select count(1) from test.rvpezocoav;


select count(1)
from rvpezocoav pezoco
join rtco co on pezoco.co_id = co.lo_id
where ad_id = 1
        and pe_id = 1
        and lo_id = 1;
```

**Comment by MathiasVersichele** [comment]
Actually, I can reproduce the unexpected behaviour. I just didn't notice because the tables are so small, so the query never errors out with the "citus.max_intermediate_result_size " message.

When using "UNION ALL" (as above), both queries below work but they show distributed query plans. This is unexpected and causes the queries with real-life data to fail.

```
explain analyze
select count(1) from test.rvpezocoav;

explain analyze 
select count(1)
from rvpezocoav pezoco
join rtco co on pezoco.co_id = co.lo_id
where ad_id = 1
        and pe_id = 1
        and lo_id = 1;
```

**Comment by MathiasVersichele** [comment]
Tried fiddling around with the view to locate what's causing the re-shuffling

The following view produces a query plan without re-shuffling:

```
CREATE OR REPLACE VIEW test.rvpezocoav as
SELECT rtpezocoav.id,
    rtpezocoav.ad_id,
    rtpezocoav.pe_id,
    rtpezocoav.zo_id,
    rtpezocoav.co_id,
    rtpezocoav.value_num,
    rtpezocoav.value_text
   FROM test.rtpezocoav
UNION ALL  --!!!!!
 SELECT av.id,
    av.ad_id,
    1 AS pe_id,
    av.zo_id,
    av.co_id,
    av.value_num,
    av.value_text
   FROM test.rtpezocoav av
     --JOIN test.rtpe ON av.pe_id = rtpe.baseline_id;

explain analyze
select count(1) from test.rvpezocoav;
```

gives

```
Aggregate  (cost=250.00..250.02 rows=1 width=8) (actual time=58.917..58.918 rows=1 loops=1)
  ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=8) (actual time=58.890..58.897 rows=48 loops=1)
        Task Count: 48
        Tuple data received from nodes: 48 bytes
        Tasks Shown: One of 48
        ->  Task
...
```

But uncommenting the line where I join on rtpe triggers the re-shuffling:

```
Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0) (actual time=0.682..0.682 rows=1 loops=1)
  ->  Distributed Subplan 53_1
        Subplan Duration: 162.17 ms
        Intermediate Data Size: 124 bytes
        Result destination: Write locally
        ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=64) (actual time=57.404..57.406 rows=2 loops=1)
              Task Count: 48
              Tuple data received from nodes: 12 bytes
              Tasks Shown: One of 48
              ->  Task
                    Tuple data received from node: 0 bytes
                    Node: host=88.99.201.54 port=50001 dbname=projects
                    ->  Seq Scan on rtpezocoav_120537 rtpezocoav  (cost=0.00..18.80 rows=880 width=64) (actual time=0.004..0.004 rows=0 loops=1)
                        Planning Time: 0.062 ms
                        Execution Time: 0.028 ms
        Planning Time: 0.000 ms
        Execution Time: 57.421 ms
  ->  Distributed Subplan 53_2
        Subplan Duration: 99.38 ms
        Intermediate Data Size: 0 bytes
        Result destination: Write locally
        ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=100000 width=64) (actual time=61.814..61.815 rows=0 loops=1)
              Task Count: 48
              Tuple data received from nodes: 0 bytes
              Tasks Shown: One of 48
              ->  Task
                    Tuple data received from node: 0 bytes
                    Node: host=88.99.201.54 port=50001 dbname=projects
                    ->  Hash Join  (cost=29.80..239.60 rows=5280 width=64) (actual time=0.020..0.022 rows=0 loops=1)
                          Hash Cond: (rtpe.baseline_id = av.pe_id)
                          ->  Seq Scan on rtpe_120585 rtpe  (cost=0.00..22.00 rows=1200 width=4) (actual time=0.009..0.010 rows=1 loops=1)
                          ->  Hash  (cost=18.80..18.80 rows=880 width=64) (actual time=0.003..0.003 rows=0 loops=1)
                                Buckets: 1024  Batches: 1  Memory Usage: 8kB
                                ->  Seq Scan on rtpezocoav_120537 av  (cost=0.00..18.80 rows=880 width=64) (actual time=0.002..0.002 rows=0 loops=1)
                        Planning Time: 0.210 ms
                        Execution Time: 0.073 ms
        Planning Time: 0.000 ms
        Execution Time: 61.837 ms
  ->  Distributed Subplan 53_3
        Subplan Duration: 0.90 ms
        Intermediate Data Size: 124 bytes
        Result destination: Write locally
        ->  Custom Scan (Citus Adaptive)  (cost=0.00..0.00 rows=0 width=0) (actual time=0.971..0.972 rows=2 loops=1)
              Task Count: 1
              Tuple data received from nodes: 12 bytes
              Tasks Shown: All
              ->  Task
                    Tuple data received from node: 12 bytes
                    Node: host=localhost port=5444 dbname=projects
                    ->  Append  (cost=0.00..0.07 rows=2 width=64) (actual time=0.073..0.127 rows=2 loops=1)
                          ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..0.04 rows=1 width=64) (actual time=0.072..0.074 rows=2 loops=1)
                          ->  Function Scan on read_intermediate_result intermediate_result_1  (cost=0.00..0.02 rows=1 width=64) (actual time=0.050..0.050 rows=0 loops=1)
                        Planning Time: 0.119 ms
                        Execution Time: 0.159 ms
        Planning Time: 0.000 ms
        Execution Time: 0.990 ms
  Task Count: 1
  Tuple data received from nodes: 1 bytes
  Tasks Shown: All
  ->  Task
        Tuple data received from node: 1 bytes
        Node: host=localhost port=5444 dbname=projects
        ->  Aggregate  (cost=0.10..0.11 rows=1 width=8) (actual time=0.072..0.073 rows=1 loops=1)
              ->  Function Scan on read_intermediate_result intermediate_result  (cost=0.00..0.09 rows=4 width=0) (actual time=0.062..0.064 rows=2 loops=1)
            Planning Time: 0.061 ms
            Execution Time: 0.104 ms
Planning Time: 3.620 ms
Execution Time: 263.215 ms
```

**Comment by onderkalaci** [comment]
This might be a Citus bug.  Still, there could be some workarounds, which you can check if works for you.


> explain analyze select count(1) from test.rvpezocoav;

The problem internally triggered by `count(1)` in the target list, if you have ` select * from test.rvpezocoav` the query can be pushed down (e.g., no data shuffle).

To prevent shuffles, you could consider adding `GROUP BY distribution_key` to the  query such as we get the count for each group, and sum them. It of course does more computation, but might work reasonably well.
```SQL
WITH cte AS
  (SELECT count(*)  as cnt,
          zo_id
   FROM
     (SELECT *
      FROM test.rvpezocoav) AS foo
   GROUP BY 2)
SELECT sum(cnt) as total_count
FROM cte;



WITH cte AS (
select count(*) as cnt, zo_id
from rvpezocoav pezoco
join rtco co on pezoco.co_id = co.lo_id
where ad_id = 1
        and pe_id = 1
        and lo_id = 1 GROUP BY zo_id)

SELECT sum(cnt) as total_count
FROM cte;
```


This works on PG12  / citus 9.4.

**Comment by MathiasVersichele** [comment]
Thx for the feedback. I can use the workaround for now, but would be good to have it bugfixed if it really is a bug.

**Comment by MathiasVersichele** [comment]
Let me know if I can close this ticket, or you want to leave it open for the bug investigation.

**Comment by onderkalaci** [comment]
Let's keep the issue open, I'll update the description with a smaller set of  steps to reproduce (e.g., we  don't need UNION to reproduce the issue)

**Comment by SaitTalhaNisanci** [comment]
This should be fixed with #5079, I can reproduce this with master but not after #5079.

**Comment by onderkalaci** [comment]
@MathiasVersichele this bug is going to be fixed on Citus 10.2
